File: apps/ios/WindmillKit/Sources/WindmillGym/Bodyweight.swift
import Foundation
import WindmillPlatform

// A weigh-in is one number on one local calendar day, and the day is its identity: a second write for the same
// `dateLocal` is a correction, decided by `recordedAt` — the newer of the two wins, on the server and on this device.
// Kilograms everywhere; the unit toggle does not convert on this phone, and the room says so in its own words.
public struct BodyweightEntry: Equatable, Codable, Sendable, Identifiable {
    public let dateLocal: String
    public let weightKg: Double
    public let recordedAt: Int64

    public init(dateLocal: String, weightKg: Double, recordedAt: Int64) {
        self.dateLocal = dateLocal
        self.weightKg = weightKg
        self.recordedAt = recordedAt
    }

    public var id: String { dateLocal }

    enum CodingKeys: String, CodingKey {
        case dateLocal, weightKg, recordedAt
    }

    public init(from decoder: Decoder) throws {
        let fields = try decoder.container(keyedBy: CodingKeys.self)
        dateLocal = try fields.decode(String.self, forKey: .dateLocal)
        weightKg = try fields.decode(Double.self, forKey: .weightKg)
        recordedAt = try fields.decodeIfPresent(Int64.self, forKey: .recordedAt) ?? 0
    }

    // Which of two writes for one day stands: the later `recordedAt`, and on a tie the one already held.
    public func supersedes(_ held: BodyweightEntry) -> Bool {
        recordedAt > held.recordedAt
    }
}

public struct BodyweightWrite: Equatable, Encodable, Sendable {
    public let weightKg: Double
    public let recordedAt: Int64

    public init(weightKg: Double, recordedAt: Int64) {
        self.weightKg = weightKg
        self.recordedAt = recordedAt
    }
}

public enum Bodyweight {
    public static let minKg = 20.0
    public static let maxKg = 400.0
    // A line joins two dots only when they are this close; a missed week is the thing worth seeing.
    public static let gapDays = 7

    public static let title = "Bodyweight"
    public static let chip = "Weigh in"
    public static let sheetTitle = "Weigh in"
    public static let save = "Save"
    public static let hint = "comma or point, both read as a decimal"
    public static let notANumber = "That is not a number yet."
    public static let oneDecimalPoint = "One decimal point only."
    public static let outOfBounds = "Between 20 and 400 kg — check the number."
    public static let notAForecast = "A weigh-in is not a forecast — today or earlier."
    public static let gapRule = "no line is drawn across a gap longer than seven days"
    public static let deleteRow = "Delete weigh-in"
    public static let deleteTitle = "Delete this weigh-in?"
    public static let deleteConfirm = "Delete"
    public static let deleteKeep = "Keep it"
    // The settings screen's own words for the unit toggle this phone does not apply.
    public static let drawsKg = "Not on this phone yet — this room still draws kg."
    public static let dateRow = "Date"

    public enum Window: String, CaseIterable, Sendable {
        case ninetyDays = "90 days"
        case all = "All"

        // Printed on the chart, beside the rule, so the window shown is never silent.
        public var printed: String {
            switch self {
            case .ninetyDays: return "last 90 days"
            case .all: return "the whole series"
            }
        }
    }

    // Two decimals at most, trailing zeros stripped: 82.4, 82.45, 82.
    public static func format(_ kg: Double) -> String {
        let rounded = (kg * 100).rounded() / 100
        guard rounded.isFinite else { return "—" }
        if rounded == rounded.rounded() { return String(Int(rounded)) }
        var text = String(format: "%.2f", rounded)
        while text.hasSuffix("0") { text.removeLast() }
        return text
    }

    // `82.4 kg · 3 days ago` for the newest weigh-in up to today; nil draws nothing — never a dash, never a zero.
    // A row dated after today (a clock behind the log's, a row another surface let through) is never the reading.
    public static func reading(_ entries: [BodyweightEntry], today: String) -> String? {
        guard let latest = latest(entries, today: today) else { return nil }
        return "\(format(latest.weightKg)) kg · \(ago(latest.dateLocal, today: today))"
    }

    public static func latest(_ entries: [BodyweightEntry], today: String) -> BodyweightEntry? {
        entries.filter { $0.dateLocal <= today }.max { $0.dateLocal < $1.dateLocal }
    }

    public static func ago(_ dateLocal: String, today: String) -> String {
        let days = daysBetween(dateLocal, today) ?? 0
        if days <= 0 { return "today" }
        if days == 1 { return "yesterday" }
        return "\(days) days ago"
    }

    // A weigh-in is what happened: a day after the device's local today is refused at the field and at the store.
    public static func dateRefusal(_ dateLocal: String, today: String) -> String? {
        guard isDateLocal(dateLocal) else { return "could not read that date" }
        return dateLocal > today ? notAForecast : nil
    }

    public struct Reading: Equatable, Sendable {
        public let value: Double?
        public let refusal: String?

        public var isValid: Bool { value != nil }
    }

    // Refusals one at a time, in the order a typist meets them; comma and point both read as the decimal, and a
    // signed number is a number — a negative one is out of bounds, not unreadable.
    public static func read(_ text: String) -> Reading {
        let raw = text.trimmingCharacters(in: .whitespacesAndNewlines)
        let normalised = raw.replacingOccurrences(of: ",", with: ".")
        guard normalised.filter({ $0 == "." }).count <= 1 else {
            return Reading(value: nil, refusal: oneDecimalPoint)
        }
        let unsigned = normalised.hasPrefix("-") || normalised.hasPrefix("+") ? String(normalised.dropFirst()) : normalised
        guard !unsigned.isEmpty, unsigned.allSatisfy({ $0.isNumber || $0 == "." }),
              let value = Double(normalised), value.isFinite else {
            return Reading(value: nil, refusal: notANumber)
        }
        let rounded = (value * 100).rounded() / 100
        guard rounded >= minKg, rounded <= maxKg else {
            return Reading(value: nil, refusal: outOfBounds)
        }
        return Reading(value: rounded, refusal: nil)
    }

    // ── the local calendar day ─────────────────────────────────────────────────────────────────

    public static func dateLocal(_ date: Date, calendar: Calendar = .current) -> String {
        let parts = calendar.dateComponents([.year, .month, .day], from: date)
        return String(format: "%04d-%02d-%02d", parts.year ?? 1970, parts.month ?? 1, parts.day ?? 1)
    }

    // A real date in `YYYY-MM-DD`, or nil: the server refuses anything else and so does the store.
    public static func date(of dateLocal: String, calendar: Calendar = .current) -> Date? {
        let pieces = dateLocal.split(separator: "-", omittingEmptySubsequences: false)
        guard pieces.count == 3, pieces[0].count == 4, pieces[1].count == 2, pieces[2].count == 2,
              let year = Int(pieces[0]), let month = Int(pieces[1]), let day = Int(pieces[2]) else { return nil }
        var parts = DateComponents()
        parts.year = year
        parts.month = month
        parts.day = day
        parts.hour = 12
        guard let made = calendar.date(from: parts) else { return nil }
        let back = calendar.dateComponents([.year, .month, .day], from: made)
        guard back.year == year, back.month == month, back.day == day else { return nil }
        return made
    }

    public static func isDateLocal(_ dateLocal: String) -> Bool {
        date(of: dateLocal) != nil
    }

    // Calendar days from `from` to `to`; nil when either is not a date.
    public static func daysBetween(_ from: String, _ to: String, calendar: Calendar = .current) -> Int? {
        guard let start = date(of: from, calendar: calendar), let end = date(of: to, calendar: calendar) else {
            return nil
        }
        return calendar.dateComponents([.day], from: calendar.startOfDay(for: start),
                                       to: calendar.startOfDay(for: end)).day
    }


    private static let months = ["Jan", "Feb", "Mar", "Apr", "May", "Jun",
                                 "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]

    // `7 Jul`, spelled rather than localised so a locale cannot reorder it.
    public static func dayMonth(_ dateLocal: String) -> String {
        let pieces = dateLocal.split(separator: "-")
        guard pieces.count == 3, let month = Int(pieces[1]), let day = Int(pieces[2]) else { return dateLocal }
        return "\(day) \(months[max(0, min(11, month - 1))])"
    }

    public static func dayMonthYear(_ dateLocal: String) -> String {
        let pieces = dateLocal.split(separator: "-")
        guard pieces.count == 3, let year = Int(pieces[0]) else { return dateLocal }
        return "\(dayMonth(dateLocal)) \(year)"
    }

    // `Tue 4 Aug`, for a day the sheet has fixed.
    public static func weekdayDayMonth(_ dateLocal: String, calendar: Calendar = .current) -> String {
        guard let day = date(of: dateLocal, calendar: calendar) else { return dateLocal }
        return Readout.day(Int64(day.timeIntervalSince1970 * 1000))
    }

    // ── the chart ──────────────────────────────────────────────────────────────────────────────

    public struct Gap: Equatable, Sendable, Identifiable {
        // The last dot before the gap and the first after it — the same two days web and Android name.
        public let from: String
        public let to: String

        public var id: String { from }
        public var label: String { "no weigh-in · \(Bodyweight.dayMonth(from)) – \(Bodyweight.dayMonth(to))" }
    }

    // A dot per measurement; `runs` are the dots a segment may join, in order; `gaps` are the spans left empty.
    public struct Chart: Equatable, Sendable {
        public let window: Window
        public let points: [BodyweightEntry]
        public let runs: [[BodyweightEntry]]
        public let gaps: [Gap]
        public let low: Double
        public let high: Double

        public var isEmpty: Bool { points.isEmpty }

        // The window and what it holds, printed on the chart: `last 90 days · 12 weigh-ins`.
        public var label: String {
            "\(window.printed) · \(points.count == 1 ? "1 weigh-in" : "\(points.count) weigh-ins")"
        }
    }

    // Ascending, never a day after today; the narrow window is the 90 calendar days ending today.
    public static func windowed(_ entries: [BodyweightEntry], window: Window, today: String) -> [BodyweightEntry] {
        let ascending = entries.filter { $0.dateLocal <= today }.sorted { $0.dateLocal < $1.dateLocal }
        guard window == .ninetyDays else { return ascending }
        return ascending.filter { (daysBetween($0.dateLocal, today) ?? 0) < 90 }
    }

    public static func chart(_ entries: [BodyweightEntry], window: Window, today: String) -> Chart {
        let points = windowed(entries, window: window, today: today)
        var runs: [[BodyweightEntry]] = []
        var gaps: [Gap] = []
        for point in points {
            guard let previous = runs.last?.last else {
                runs.append([point])
                continue
            }
            let apart = daysBetween(previous.dateLocal, point.dateLocal) ?? 0
            if apart <= gapDays {
                runs[runs.count - 1].append(point)
                continue
            }
            gaps.append(Gap(from: previous.dateLocal, to: point.dateLocal))
            runs.append([point])
        }
        let weights = points.map(\.weightKg)
        let least = weights.min() ?? 0
        let most = weights.max() ?? 0
        // The series' own range plus padding, so a kilo of movement is visible and a flat series is not a wall.
        let padding = max(0.5, (most - least) * 0.15)
        return Chart(window: window, points: points, runs: runs, gaps: gaps,
                     low: least - padding, high: most + padding)
    }

    // What the chart screen says when the window holds nothing; nil while it holds a dot.
    public static func emptyWindow(_ chart: Chart) -> String? {
        guard chart.isEmpty else { return nil }
        return chart.window == .ninetyDays ? "no weigh-in in the last 90 days" : "no weigh-in yet"
    }
}

File: apps/android/gym/src/main/kotlin/works/windmill/gym/domain/Bodyweight.kt
package works.windmill.gym.domain

import java.math.BigDecimal
import java.math.RoundingMode
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.temporal.ChronoUnit
import kotlin.math.ceil
import kotlin.math.floor
import kotlinx.serialization.Serializable

// One row per local calendar date; the date IS the identity, so every write is idempotent by it and
// the server keeps whichever of two writes carries the later `recordedAt`. Kilograms only, on the
// wire and on this phone. `recordedAt` is this device's clock at the save and decides nothing but
// which of two writes is newer.
@Serializable
data class WeighIn(val dateLocal: String, val weightKg: Double, val recordedAt: Long) {
    val date: LocalDate get() = LocalDate.parse(dateLocal)
}

// PUT /v1/gym/bodyweight/{dateLocal}. Both fields ride explicitly: encodeDefaults is off and neither
// has a default to omit.
@Serializable
data class WeighInWrite(val weightKg: Double, val recordedAt: Long)

// What the field said, or the one refusal it met first.
sealed interface ParsedWeight {
    data class Ok(val weightKg: Double) : ParsedWeight
    data class Refused(val said: String) : ParsedWeight
}

// A dot per measurement; a segment joins two dots only when the gap between them is short enough to
// be an ordinary week. A longer gap is left empty and named.
sealed interface ChartRun {
    data class Segment(val from: WeighIn, val to: WeighIn) : ChartRun
    data class Gap(val from: WeighIn, val to: WeighIn) : ChartRun {
        val label: String get() = "no weigh-in · ${Bodyweight.shortDay(from.date)} – ${Bodyweight.shortDay(to.date)}"
    }
}

enum class ChartWindow(val label: String) {
    Ninety("90 days"), All("All");
}

object Bodyweight {
    const val title = "Bodyweight"
    const val chip = "Weigh in"
    const val save = "Save"
    const val fieldHint = "comma or point, both read as a decimal"
    const val unit = "kg"

    // The refusals, one at a time, in the order the field is read.
    const val notANumber = "That is not a number yet."
    const val onePoint = "One decimal point only."
    const val outOfRange = "Between 20 and 400 kg — check the number."
    const val notAForecast = "A weigh-in is not a forecast — today or earlier."

    const val deleteRow = "Delete weigh-in"
    const val deleteAsk = "Delete this weigh-in?"
    const val delete = "Delete"

    // Printed on the chart, because a segment is a connection and not data.
    const val gapRule = "no line is drawn across a gap longer than seven days"
    const val maxGapDays = 7L

    const val nothingYet = "No weigh-ins yet. Weigh in from the log."
    const val noneInWindow = "no weigh-in in the last 90 days"

    // The same words the settings screen uses: this phone converts nothing.
    const val kilogramsOnly = "This phone still draws every weight in kilograms — nothing on this screen converts one."

    const val minKg = 20.0
    const val maxKg = 400.0

    fun today(nowMs: Long): LocalDate =
        Instant.ofEpochMilli(nowMs).atZone(ZoneId.systemDefault()).toLocalDate()

    // Comma or point, at most one of either, digits on at least one side, and inside the bounds.
    fun parse(typed: String): ParsedWeight {
        val raw = typed.trim()
        if (raw.isEmpty() || raw.none { it.isDigit() }) return ParsedWeight.Refused(notANumber)
        if (raw.count { it == '.' || it == ',' } > 1) return ParsedWeight.Refused(onePoint)
        val normalised = raw.replace(',', '.')
        if (!normalised.all { it.isDigit() || it == '.' }) return ParsedWeight.Refused(notANumber)
        val value = normalised.toBigDecimalOrNull() ?: return ParsedWeight.Refused(notANumber)
        val rounded = value.setScale(2, RoundingMode.HALF_UP).toDouble()
        if (rounded < minKg || rounded > maxKg) return ParsedWeight.Refused(outOfRange)
        return ParsedWeight.Ok(rounded)
    }

    // A date past this device's local today is refused at the field; the log refuses one more than a
    // day past its own UTC today in the same words.
    fun dated(date: LocalDate, today: LocalDate): String? =
        if (date.isAfter(today)) notAForecast else null

    // Two decimals stored; trailing zeros are not drawn: 82.40 reads 82.4, 82.00 reads 82.
    fun kilograms(weightKg: Double): String =
        BigDecimal.valueOf(weightKg).setScale(2, RoundingMode.HALF_UP).stripTrailingZeros().toPlainString()

    // The y-axis names its unit: `85.5 kg`.
    fun axisLabel(weightKg: Double): String = "${kilograms(weightKg)} $unit"

    // `82.4 kg · 3 days ago`. A calendar claim: both ends are local dates.
    fun reading(latest: WeighIn?, nowMs: Long): String? {
        val held = latest ?: return null
        return "${kilograms(held.weightKg)} $unit · ${age(held.date, today(nowMs))}"
    }

    fun age(date: LocalDate, today: LocalDate): String {
        val days = ChronoUnit.DAYS.between(date, today)
        if (days <= 0) return "today"
        if (days == 1L) return "yesterday"
        return "$days days ago"
    }

    private val months = listOf("Jan", "Feb", "Mar", "Apr", "May", "Jun",
                                "Jul", "Aug", "Sep", "Oct", "Nov", "Dec")
    private val weekdays = listOf("Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat")

    fun shortDay(date: LocalDate): String = "${date.dayOfMonth} ${months[date.monthValue - 1]}"

    // `Tue 26 Aug`, or `Today · Tue 26 Aug` when it is.
    fun dayLine(date: LocalDate, today: LocalDate): String {
        val day = "${weekdays[date.dayOfWeek.value % 7]} ${shortDay(date)}"
        if (date == today) return "Today · $day"
        return day
    }

    // The reading: the newest day that has happened. A served row dated after this device's today is
    // never the reading and never a dot — the log refused it a day later than this phone would.
    fun latest(entries: List<WeighIn>, today: LocalDate): WeighIn? =
        entries.filter { !it.date.isAfter(today) }.maxByOrNull { it.dateLocal }

    // Ascending by date, never past today; the last 90 calendar days ending today for the narrow window.
    fun windowed(entries: List<WeighIn>, window: ChartWindow, today: LocalDate): List<WeighIn> {
        val sorted = entries.filter { !it.date.isAfter(today) }.sortedBy { it.dateLocal }
        if (window == ChartWindow.All) return sorted
        val from = today.minusDays(89)
        return sorted.filter { !it.date.isBefore(from) }
    }

    // Printed on the chart: `last 90 days · 4 weigh-ins`, the same words on every surface.
    fun windowLine(window: ChartWindow, shown: Int): String {
        val counted = if (shown == 1) "1 weigh-in" else "$shown weigh-ins"
        return when (window) {
            ChartWindow.Ninety -> "last 90 days · $counted"
            ChartWindow.All -> "the whole series · $counted"
        }
    }

    // The repair sheet is titled by the day it repairs; the entry sheet by the verb.
    fun sheetTitle(fixedDate: LocalDate?): String =
        if (fixedDate == null) chip else "Weigh-in · ${shortDay(fixedDate)}"

    // Consecutive dots, joined or left apart on the seven-day rule.
    fun runs(entries: List<WeighIn>): List<ChartRun> =
        entries.sortedBy { it.dateLocal }.zipWithNext { from, to ->
            if (ChronoUnit.DAYS.between(from.date, to.date) > maxGapDays) ChartRun.Gap(from, to)
            else ChartRun.Segment(from, to)
        }

    // The y-axis is the series' own floor and ceiling plus padding, never zero: a bodyweight between
    // 82.0 and 84.5 has to be readable as a slope.
    data class Axis(val floorKg: Double, val ceilingKg: Double) {
        fun fraction(weightKg: Double): Float {
            val span = ceilingKg - floorKg
            if (span <= 0.0) return 0.5f
            return ((weightKg - floorKg) / span).toFloat().coerceIn(0f, 1f)
        }
    }

    fun axis(entries: List<WeighIn>): Axis? {
        if (entries.isEmpty()) return null
        val low = entries.minOf { it.weightKg }
        val high = entries.maxOf { it.weightKg }
        val padding = maxOf(1.0, (high - low) * 0.2)
        return Axis(
            floorKg = floor((low - padding) * 2) / 2,
            ceilingKg = ceil((high + padding) * 2) / 2,
        )
    }
}
